### PR TITLE
Fix #286 reset sync poison CKSyncEngine state

### PR DESCRIPTION
**Pull request by mnbf9rca**
Fixes #286

## Summary
- Discards a restored CKSyncEngine serialization when it is reset-poisoned by a pending DeviceSync zone deletion or reset-in-flight state, then re-bootstraps from durable reset/store intent.
- Moves CKSyncEngine `fetchChanges()` and `sendChanges()` calls behind a `Task.detached` boundary at `CKSyncEngineDriver`, matching the confirmed iOS assert mitigation.
- Keeps reset-path hygiene that clears pending engine changes before destructive zone reset sends, while preserving #296 facade readiness gating and #300 snapshot delete behavior.

## Centerpiece Evidence
Confirmed pre-fix device assert from Console.app:

`CloudKit/CKSyncEngine.swift:318: Fatal error: BUG IN CLIENT OF CLOUDKIT: Cannot await a call into CKSyncEngine from within a delegate callback if that function will end up calling back into the delegate. Otherwise, CKSyncEngine cannot guarantee that it calls your delegate callbacks serially. Try performing this in a detached Task.`

Research trail: `docs/plans/286-cksyncengine-assert-research-synthesis.md`.

## #286 x #296 Composition Check
- Self-heal runs in `SyncEngineController.start()` immediately after driver creation and before startup completion / `ProfileSyncManager.markSyncReadyAndFlush()`, so a poisoned launch cannot become facade-ready mid-heal.
- Reset send path still drives the controller/driver directly via `DriverResetOutbox.requestSend()`; it does not depend on #296 facade enqueue gating. The production driver now detaches before awaiting CKSyncEngine.
- Spot-check audit preserved merged code from #289/#294/#296/#300: `scheduleProfileDeleteCommit`, `onPendingDeleteEnqueued`, `enqueueTombstoneDelete`, and `BlockedProfileCardData` are present; `MutationFunnel`, `BlockedProfiles`, and the card family have no PR diff.

## Bring-Forward Diff Audit
`git diff origin/main..origin/fix/286-reset-sync-poison-signed --stat`:

```
 Foqos/CloudKit/ProfileSyncManager.swift            |  1 +
 Foqos/CloudKit/SyncEngine/CKSyncEngineDriver.swift | 19 ++++-
 Foqos/CloudKit/SyncEngine/ResetController.swift    |  8 +-
 .../SyncEngine/SyncEngineController+Cutover.swift  |  1 +
 .../SyncEngine/SyncEngineController+Reset.swift    | 11 ++-
 .../CloudKit/SyncEngine/SyncEngineController.swift | 25 +++++-
 Foqos/CloudKit/SyncEngine/SyncEngineDriver.swift   |  4 +-
 FoqosTests/CKSyncEngineTaskBoundaryTests.swift     | 35 +++++++++
 FoqosTests/Mocks/ResetSeamMocks.swift              |  2 +
 FoqosTests/SyncEngineControllerTests.swift         | 52 +++++++++++++
 FoqosTests/SyncEngineResetTests.swift              | 51 +++++++++++++
 docs/sync-engine-two-device-checklist.md           | 89 +++++++++++++++++-----
 12 files changed, 273 insertions(+), 25 deletions(-)
```

Exact tree comparison between the locally verified branch and signed replay branch passed: `git diff --exit-code HEAD origin/fix/286-reset-sync-poison-signed --stat`.

Signed replay commit: aaf4f12ffec016a97a3d5e42c1445b6c25606cba, GitHub verification `verified: true`.

## Verification
- `swift-format lint --recursive .` — pass.
- Focused integration XCTest set (#286 reset/controller + #296 facade/attach + #300 snapshot/card): 107 tests, 0 failures.
- Full XCTest suite: 850 passed, 0 failed, 0 skipped on simulator UUID `B9E4A679-BDF3-4541-A59F-DA4BE21F80ED`, `-parallel-testing-enabled NO`.
- `scripts/check-c2-guards.sh` — pass.
- `scripts/check-sync-guards.sh` — pass.

Existing unrelated warning observed during build: `HeartbeatManager.swift:42` missing `.approvedWithDataAccess` exhaustiveness case.

## Two-Device Checklist
| Scenario | Result | Evidence / Notes |
| --- | --- | --- |
| Reset Sync — Keep App Selections | PASS | Maintainer-confirmed no crash after detached fix; logs show reset delete/save/reseed path completing. |
| Reset Sync — Clear App Selections | PASS | Maintainer-confirmed no crash after detached fix; logs show reset delete/save/reseed path completing. |
| Concurrent edit | PASS | Maintainer result: pass. |
| Concurrent delete | PASS (extra) | Maintainer result: pass. |
| Device offline / airplane across reset | PASS with residual | No crash/assert; offline/reset recovery exercised. Local registration gap tracked separately as #301. |
| Poisoned-install self-heal | PASS | Install over poisoned/crash-looping state recovered on launch; log contains `[#286] restored engine state carries a pending zone-deletion; discarding serialization and re-bootstrapping`. |
| Toggle off -> local delete -> on | FAIL, known issue #302 | Disabled-sync local deletes bypass funnel tombstone bookkeeping; this row remains a contract gap and is not #286 scope. Evidence: `docs/plans/FamilyFoqos-Logs-2026-07-10-170124.log`. |
| Toggle off -> remote delete -> on | PASS | Rejoin seed resurrected remote-deleted record as N5 keep-bias expects; log shows re-enable seed and `unknownItem` re-add for the missing record. Evidence: `docs/plans/FamilyFoqos-Logs-2026-07-10-171306.log`. |
| Optional stress / not-routine rows | Not routine / not executed unless noted above | Checklist now documents hard-to-force rows and recovery expectations; #302 is the only known failed required row. |

## Deviations
- The temporary pending-queue instrumentation used to falsify the queue-shape hypothesis was removed before PR. Low-volume sync lifecycle logging remains.
- Checklist row `Toggle off -> local delete -> on` is recorded as failed known issue #302, not fixed in #286.
- No queue-shape choke-point invariant or reset-window send gate was added; research and device assert evidence showed the root cause was CKSyncEngine delegate task-local inheritance, fixed at the driver boundary with `Task.detached`.

## Summary by Sourcery

Handle CKSyncEngine reset poisoning by discarding unsafe restored engine state and enforcing detached task boundaries around CloudKit sync operations, while tightening reset hygiene and logging around zone deletions and sync readiness.

Bug Fixes:
- Prevent crashes/asserts by running CKSyncEngine fetch/send operations across a detached task boundary instead of within delegate callbacks.
- Ensure reset paths clear pending engine changes before issuing destructive zone deletes to avoid invalid change combinations.
- Discard restored CKSyncEngine serializations that indicate in-flight reset or pending zone deletion, then rebuild a fresh engine driven by durable reset/store intent.

Enhancements:
- Add defensive checks in SyncEngineController to detect and self-heal poisoned engine state on startup without disrupting existing reset/facade behavior.
- Improve reset and sync logging for zone delete/save operations, reset command sends, and sync readiness transitions.
- Refine the manual two-device sync acceptance checklist to distinguish required scenarios from optional stress and non-routine cases, clarifying expected recovery via Reset Sync.

Tests:
- Add unit tests covering detached task boundary behavior for CKSyncEngine operations and ensuring TaskLocal context is not inherited.
- Add tests validating reset coexistence guards, including clearing pending saves before zone delete and resumed deleting stages not mixing deleteZone with saves.
- Add tests verifying startup self-heal behavior for poisoned vs healthy restored engine state in the SyncEngineController.

<h3>Greptile Summary</h3>

This PR fixes the CKSyncEngine delegate-reentrancy fatal assert (`BUG IN CLIENT OF CLOUDKIT`) by wrapping `fetchChanges()` and `sendChanges()` in `Task.detached` via a new `CKSyncEngineTaskBoundary` helper, and adds defensive hygiene that discards a restored engine serialization carrying a pending zone-deletion (or any active reset intent) to prevent a "poison" state from persisting across app restarts.

- **Core assert fix**: `CKSyncEngineTaskBoundary.runDetached` wraps both driver calls in `Task.detached`, breaking task-local inheritance from the delegate-callback chain; the task-boundary test verifies this isolation property.
- **Self-heal on start**: `restoredStateIsPoisoned()` detects a pending `deleteZone` in the restored serialization or any active `resetIntent`, discards the serialization (accepting a full re-fetch as the recovery cost), and rebuilds a fresh driver before the startup sequence runs.
- **Queue hygiene**: `clearPendingChangesForReset()` purges all pending record and database changes before the zone-delete send in `beginReset()` and `reenqueueDeleting()`, preventing coexisting record-saves alongside a zone delete that CKSyncEngine would assert on.

<h3>Confidence Score: 4/5</h3>

The fatal CKSyncEngine assert fix is correct and the two-device checklist confirms no crashes after the change. The defensive self-heal and queue hygiene are conservative but sound. The main open question is whether discarding the serialization during a mid-seeding restart is worth the forced full re-fetch.

The `Task.detached` boundary in `fetchChanges()`/`sendChanges()` directly addresses the confirmed device assert and is validated by the task-local isolation test. The poison check and `clearPendingChangesForReset()` add correctness guardrails with good test coverage. Two observations hold the score below the maximum: the poison check fires for all `resetIntent` stages including `.seeding` (where the zone is healthy), unnecessarily discarding valid fetch tokens; and `applySeedDecision()` can enqueue a redundant seed pass when a reset resume also triggers seeding, relying on CKSyncEngine deduplication to avoid duplicate sends. Both are functional but inefficient.

`SyncEngineController.swift` — the `restoredStateIsPoisoned()` guard and its interaction with `applySeedDecision()` in the startup sequence are the most complex parts of the change and worth a second look, particularly for `.seeding`-stage restarts.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| Foqos/CloudKit/SyncEngine/CKSyncEngineDriver.swift | Adds `CKSyncEngineTaskBoundary.runDetached` and wraps `fetchChanges()`/`sendChanges()` with it; directly fixes the CKSyncEngine delegate-reentrancy fatal assert. |
| Foqos/CloudKit/SyncEngine/SyncEngineController.swift | Adds `restoredStateIsPoisoned()` self-heal check in `start()`; triggers when any `resetIntent` is active (not only when the serialization carries a pending deleteZone), which discards valid fetch tokens and forces full re-fetch during mid-seeding restarts. |
| Foqos/CloudKit/SyncEngine/SyncEngineController+Reset.swift | Adds `clearPendingChangesForReset()` which removes ALL pending record zone and database changes before a zone delete; also schedules `sendChanges()` via a regular main-actor Task (isolation comes from inside `sendChanges()` via `Task.detached`). |
| Foqos/CloudKit/SyncEngine/ResetController.swift | Adds `clearPendingChangesForReset()` to `ResetOutbox` protocol and calls it in `beginReset()` and `reenqueueDeleting()` to prevent coexisting record-saves alongside a zone delete. |
| FoqosTests/CKSyncEngineTaskBoundaryTests.swift | New test verifying that `Task.detached` does not inherit a set `@TaskLocal`; accurately validates the isolation property the fix relies on. |
| FoqosTests/SyncEngineResetTests.swift | Adds two coexistence-guard tests verifying that `beginReset` and `resume(.deleting)` leave only a pending `deleteZone` with no accompanying record saves. |
| FoqosTests/SyncEngineControllerTests.swift | Adds two self-heal tests: one for discarding a poisoned serialization (pending deleteZone) and one asserting a healthy serialization is preserved. |

<sub>Reviews (1): Last reviewed commit: ["fix(#286): recover reset-poisoned sync e..."](https://github.com/mnbf9rca/family-foqos/commit/aaf4f12ffec016a97a3d5e42c1445b6c25606cba) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43335754)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

**Comment by mnbf9rca** [comment]
Ready for review. Please review before merge; this PR intentionally leaves the disabled-sync local-delete checklist row as known issue #302 and does not broaden #286 scope.

**Comment by mnbf9rca** [comment]
Re: general feedback from @sourcery-ai

**"`SyncEngineController.restoredStateIsPoisoned()` accesses `pendingDatabaseChanges` on a protocol that does not declare it"** — Won’t change. Checked the seam: `SyncEngineDriver.swift:21` declares `var pendingDatabaseChanges: [CKSyncEngine.PendingDatabaseChange] { get }`, and both production/test drivers implement it. The claim is factually incorrect.

**"Avoid constructing a driver from a reset-poisoned serialization when `resetIntent` is active"** — Accepted via the inline thread. I am changing startup so active reset discards the serialization before first driver construction, while keeping the non-reset pending-deleteZone inspection path.

File: Foqos/CloudKit/SyncEngine/SyncEngineController.swift [hunk 1]
```diff
@@ -219,6 +231,17 @@ final class SyncEngineController: SyncEngineDriverDelegate {
...
+  /// #286: a restored serialization is unsafe to keep if it carries a pending `.deleteZone`
+  /// for the sync zone, or if a `resetIntent` is in progress (the reset state machine, not
+  /// the restored queue/tokens, is the source of truth for zone changes — defense in depth).
+  private func restoredStateIsPoisoned() -> Bool {
+    if store.resetIntent != nil { return true }
+    return driver.pendingDatabaseChanges.contains {
+      if case .deleteZone(let id) = $0 { return id == zoneID }
+      return false
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/CloudKit/SyncEngine/SyncEngineController.swift:245]: Won’t change this. The approved #286 contract treats an active `resetIntent` as the recovery boundary: reset resume owns the remaining stages, and losing restored fetch tokens/full-refetch is the accepted cost of avoiding reset-poisoned CKSyncEngine state. Optimizing `.seeding` to retain a restored serialization would reintroduce stage-specific cleverness that the plan deliberately avoided; see the reset design’s §8.1/N14 crash-resume semantics.

File: Foqos/CloudKit/SyncEngine/SyncEngineController.swift [hunk 2]
```diff
@@ -104,6 +104,18 @@ final class SyncEngineController: SyncEngineDriverDelegate {
   func start() {
     guard state == .disabled || state == .purged else { return }
     driver = driverFactory(store.engineState)
+    // #286 self-heal: a serialization that carries a pending zone-deletion (or was captured
+    // mid-reset) is reset poison. Discard it and rebuild a fresh engine; resetIntent /
+    // pendingSeedIntent / tombstones live in `store` (not the serialization) and drive a
+    // clean re-seed. Lost fetch tokens (=> full re-fetch) are the accepted cost of recovering
+    // an otherwise-bricked install.
+    if store.engineState != nil && restoredStateIsPoisoned() {
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/CloudKit/SyncEngine/SyncEngineController.swift:115]: Fixed. Active `resetIntent` now clears `store.engineState` before driver construction, so startup builds the first driver with `nil` serialization instead of constructing a disposable restored engine. The non-reset pending `.deleteZone` poison inspection path is still retained. (`3b8cb598`)

File: Foqos/CloudKit/SyncEngine/SyncEngineController.swift [hunk 3]
```diff
@@ -104,6 +104,18 @@ final class SyncEngineController: SyncEngineDriverDelegate {
...
+      Log.warning(
+        "[#286] restored engine state carries a pending zone-deletion; discarding "
+          + "serialization and re-bootstrapping", category: .sync)
+      store.engineState = nil
+      driver = driverFactory(nil)
+    }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/CloudKit/SyncEngine/SyncEngineController.swift:121]: Valid concern. I did not change the reset stage semantics; instead startup now skips the generic `applySeedDecision()` while `store.resetIntent != nil`, so reset resume is the sole owner for `.deleting`, `.recreating`, and `.seeding`. Added reset-stage startup tests covering the duplicate-seed cases. (`3b8cb598`)

File: Foqos/CloudKit/SyncEngine/SyncEngineController+Reset.swift [hunk 4]
```diff
@@ -18,25 +18,34 @@ final class DriverResetOutbox: ResetOutbox {
...
   }
   func removeCommandSave() {
     driver.remove(pendingRecordZoneChanges: [.saveRecord(commandRecordID)])
   }
   func requestSend() {
-    // §1.1: schedule sendChanges() in a Task AFTER the current handler returns.
+    // §1.1: the driver uses Task.detached before awaiting CKSyncEngine.
+    Log.debug("Reset sync: request send", category: .sync)
     Task { @MainActor in self.driver.sendChanges() }
   }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/CloudKit/SyncEngine/SyncEngineController+Reset.swift:44]: Fixed. The comment now distinguishes the outer `Task { @MainActor }` as a deferral to a later main-actor turn from the actual §1.1/task-local CKSyncEngine boundary inside the driver, where `sendChanges()` crosses `Task.detached` before awaiting. (`3b8cb598`)

**Comment by mnbf9rca** [comment]
Re: general feedback from @sourcery-ai after review-round changes

**"`pendingDatabaseChanges` is missing from the protocol"** — Rejected. `SyncEngineDriver.swift:21` declares it, and the production/test drivers implement it.

**"Avoid constructing from reset-poisoned serialization when `resetIntent` is active"** — Fixed in `3b8cb598`. Active reset now discards restored serialization before the first driver construction; the non-reset pending `.deleteZone` self-heal inspection remains intact.

**Comment by mnbf9rca** [comment]
Review round 1 applied in signed commit `3b8cb59872394b925f77f41b269738848c765d44`.

Verification after the review fixes:
- `swift-format lint --recursive .` passed
- `SyncEngineControllerTests`: 66 passed, 0 failed (`B9E4A679-BDF3-4541-A59F-DA4BE21F80ED`, `-parallel-testing-enabled NO`)
- Full unit suite: 854 passed, 0 failed; Xcode still skips `FoqosUITests` with `UITargetAppPath should be provided`
- `scripts/check-c2-guards.sh` passed
- `scripts/check-sync-guards.sh` passed

Remaining merge gate from Fable/maintainer: one-pass device rerun of the two reset rows (`Reset Sync — Keep App Selections` and `Reset Sync — Delete App Selections`) on this PR head, debugger detached.

**Comment by mnbf9rca** [comment]
**Final gate: approve — merge.** Independent full suite on the final head: **854/854 passed, 0 failures** (from my own run's .xcresult). Static checks all pass: `Task.detached` correctly installed behind the named `CKSyncEngineTaskBoundary.runDetached` helper at both driver call sites (the old `Task { try? await engine.… }` lines are exactly what the diff deletes, along with the falsified §1.1 comments — every deletion inspected and legitimate); queue instrumentation fully removed (only the intentional self-heal log line remains); review-fix #4 (reset resume owns reset-stage seeding) landed with tests. Device evidence complete: full two-device checklist, the #302-annotated FAIL row correctly recorded, plus the maintainer's three-pass re-run on this exact head (both reset variants + kill-mid-reset resume: pass/pass/pass) — and the root cause is device-confirmed with the verbatim iOS 27 Beta 2 assert text.

Eight days from bricked phone to merge: three falsified queue-shape theories, one botched integration caught by the diff audit, one binary-forensics workflow that read the answer out of CloudKit's own assert string, and a two-line fix. Closes the oldest critical on the board.
